browser/psst/psst_tab_web_contents_observer_browsertest.cc

```diff
@@ -0,0 +1,252 @@
+  }
+  {
+    std::u16string expected_title(u"OK");
+    content::TitleWatcher watcher(web_contents(), expected_title);
+    ASSERT_TRUE(content::NavigateToURL(web_contents(), url));
+    EXPECT_EQ(expected_title, watcher.WaitAndGetTitle());
+    EXPECT_EQ(
+        url.spec(),
+        content::EvalJs(web_contents(), kGetCurrentUrlScript).ExtractString());
+  }
+}
+
+IN_PROC_BROWSER_TEST_F(PsstTabWebContentsObserverBrowserTest,
+                       DontStartScriptHandlerIfPsstDisabled) {
+  GetPrefs()->SetBoolean(prefs::kPsstEnabled, false);
+  EXPECT_EQ(GetPrefs()->GetBoolean(prefs::kPsstEnabled), false);
+  const GURL url = GetEmbeddedTestServer().GetURL("a.com", "/simple.html");
+
+  std::u16string expected_title(u"OK");
+  content::TitleWatcher watcher(web_contents(), expected_title);
+  ASSERT_TRUE(content::NavigateToURL(web_contents(), url));
+  EXPECT_EQ(expected_title, watcher.WaitAndGetTitle());
+  EXPECT_EQ(
+      url.spec(),
+      content::EvalJs(web_contents(), kGetCurrentUrlScript).ExtractString());
+}
+
+IN_PROC_BROWSER_TEST_F(PsstTabWebContentsObserverBrowserTest,
+                       StartScriptHandlerOnlyInPrimaryMainFrame) {
+  GetPrefs()->SetBoolean(prefs::kPsstEnabled, true);
+  EXPECT_EQ(GetPrefs()->GetBoolean(prefs::kPsstEnabled), true);
+  const GURL ifrave_url =
+      GetEmbeddedTestServer().GetURL("a.com", "/simple.html");
+  const GURL navigate_url = url(
+      ifrave_url, GetEmbeddedTestServer().GetURL("a.com", "/iframe_load.html"));
+  std::u16string expected_title(u"iframe test");
+  content::TitleWatcher watcher(web_contents(), expected_title);
+  ASSERT_TRUE(content::NavigateToURL(web_contents(), navigate_url));
+  EXPECT_EQ(expected_title, watcher.WaitAndGetTitle());
+}
+
+IN_PROC_BROWSER_TEST_F(PsstTabWebContentsObserverBrowserTest,
+                       StartScriptHandlerBothScriptsExecuted) {
+  GetPrefs()->SetBoolean(prefs::kPsstEnabled, true);
+  EXPECT_EQ(GetPrefs()->GetBoolean(prefs::kPsstEnabled), true);
+  const GURL url = GetEmbeddedTestServer().GetURL("a.com", "/simple.html");
+
+  std::u16string expected_title(u"a_user-a_policy");
+  content::TitleWatcher watcher(web_contents(), expected_title);
+  ASSERT_TRUE(content::NavigateToURL(web_contents(), url));
+  EXPECT_EQ(expected_title, watcher.WaitAndGetTitle());
+}
+
+IN_PROC_BROWSER_TEST_F(PsstTabWebContentsObserverBrowserTest,
+                       PsstPrefsNotExistBothScriptsExecuted) {
+  EXPECT_EQ(GetPrefs()->GetBoolean(prefs::kPsstEnabled), true);
+  const GURL url = GetEmbeddedTestServer().GetURL("a.com", "/simple.html");
+
+  std::u16string expected_title(u"a_user-a_policy");
+  content::TitleWatcher watcher(web_contents(), expected_title);
+  ASSERT_TRUE(content::NavigateToURL(web_contents(), url));
+  EXPECT_EQ(expected_title, watcher.WaitAndGetTitle());
+}
+
+IN_PROC_BROWSER_TEST_F(PsstTabWebContentsObserverBrowserTest,
+                       PsstPrefsNotExistFeatureDisabledNothingExecuted) {
```

Comment:
This is not sufficient, you also need to test that it does nothing when the pref is set to true (feature enabled, pref set to true and then disabled again)

Reply:
also see https://github.com/brave/brave-core/pull/29044/files#r2130474303

Reply:
for the case where the feature is disabled it's sufficient to test that PsstTabWebContentsObserver is not created. If the check is moved into `MaybeCreateForWebContents` then you can unit test this instead